avro4s-core/src/main/scala/com/sksamuel/avro4s/SchemaFor.scala

```diff
@@ -217,7 +218,7 @@ object SchemaFor {
     }
 
     // if our default value is null, then we should change the type to be nullable even if we didn't use option
-    val schemaWithPossibleNull = if (encodedDefault == JsonProperties.NULL_VALUE) {
+    val schemaWithPossibleNull = if (default == Some(null)) {
```

Comment:
Maybe this should just be `if (default == Some(null) && schema.getType != Schema.Type.Union`
Because the use case we are trying to catch is that the user has done `case class Foo(a: String = null)`. If they used None as the default, then they must have an Option on the left hand side of that, so the union will be generated anyway.

Reply:
Good shout, updated.